nvflare/fuel/f3/cellnet/cell.py

```diff
@@ -0,0 +1,1757 @@
+
+    APP_ID = 1
+    ERR_TYPE_MSG_TOO_BIG = "MsgTooBig"
+    ERR_TYPE_COMM = "CommErr"
+
+    def __init__(
+        self,
+        fqcn: str,
+        root_url: str,
+        secure: bool,
+        credentials: dict,
+        create_internal_listener: bool = False,
+        parent_url: str = None,
+        max_timeout=3600,
+        bulk_check_interval=0.5,
+        bulk_process_interval=0.5,
+        max_bulk_size=100,
+    ):
+        """
+
+        Args:
+            fqcn: the Cell's FQCN (Fully Qualified Cell Name)
+            credentials: credentials for secure connections
+            root_url: the URL for backbone external connection
+            secure: secure mode or not
+            max_timeout: default timeout for send_and_receive
+            create_internal_listener: whether to create an internal listener for child cells
+            parent_url: url for connecting to parent cell
+
+        FQCN is the names of all ancestor, concatenated with dots.
+        Note: internal listener is automatically created for root cells.
+
+        Example:
+            server.J12345       (the cell for job J12345 on the server)
+            server              (the root cell of server)
+            nih_1.J12345        (the cell for job J12345 on client_1's site)
+            client_1.J12345.R0  (the cell for rank R0 of J12345 on client_1 site)
+            client_1            (he root cell of client_1)
+
+        """
+        comm_configurator = CommConfigurator()
+        self._name = self.__class__.__name__
+        self.logger = logging.getLogger(self._name)
+        self.max_msg_size = comm_configurator.get_max_message_size()
+        self.comm_configurator = comm_configurator
+
+        err = FQCN.validate(fqcn)
+        if err:
+            raise ValueError(f"Invalid FQCN '{fqcn}': {err}")
+
+        self.my_info = FqcnInfo(FQCN.normalize(fqcn))
+        self.secure = secure
+        self.logger.debug(f"{self.my_info.fqcn}: max_msg_size={self.max_msg_size}")
+
+        if not root_url:
+            raise ValueError(f"{self.my_info.fqcn}: root_url not provided")
+
+        if self.my_info.is_root and self.my_info.is_on_server:
+            if isinstance(root_url, list):
+                for url in root_url:
+                    if not _validate_url(url):
+                        raise ValueError(f"{self.my_info.fqcn}: invalid Root URL '{url}'")
+            else:
+                if not _validate_url(root_url):
+                    raise ValueError(f"{self.my_info.fqcn}: invalid Root URL '{root_url}'")
+                root_url = [root_url]
+        else:
+            if isinstance(root_url, list):
+                # multiple urls are available - randomly pick one
+                root_url = random.choice(root_url)
+                self.logger.info(f"{self.my_info.fqcn}: use Root URL {root_url}")
+            if not _validate_url(root_url):
+                raise ValueError(f"{self.my_info.fqcn}: invalid Root URL '{root_url}'")
```

Comment:
probably better to extract this into a separate method, instead of directly write inside the constructor.

Reply:
It's just a huge constructor, no logic place to break it apart. It will take more refactoring to make the constructor small. Let's do that later.